### DISCUSSION TITLE
Can I use Email Provider without an adapter?

### DISCUSSION THREAD

**Discussion by amirhhashemi** [nextauthjs/next-auth#4047] [Help]
I've set up an [`GitHubProvider`](https://next-auth.js.org/providers/github) without using an adapter (something like [this](https://github.com/calcom/cal.com/blob/ac0c3bdfb9cdf7a8007c0750dd4752d9ab68cfa0/apps/web/pages/api/auth/%5B...nextauth%5D.tsx#L143)). I handle database stuff in the `signIn` callback and everything looks good. Now I want to add support for passwordless login with [`EmailProvider`](https://next-auth.js.org/configuration/providers/email) which requires an adapter. My current database structure is nowhere near the one that adapters expect so not only  [`EmailProvider`](https://next-auth.js.org/configuration/providers/email) doesn't work, I get this kind of errors when I want to login with GitHub:
```
https://next-auth.js.org/errors#oauth_callback_handler_error Cannot read property 'findUnique' of undefined TypeError: Cannot read property 'findUnique' of undefined
    at getUserByAccount (/home/amir/Documents/code/reflog/node_modules/@next-auth/prisma-adapter/dist/index.js:11:45)
    at _callee2$ (/home/amir/Documents/code/reflog/node_modules/next-auth/core/errors.js:315:29)
    at tryCatch (/home/amir/Documents/code/reflog/node_modules/regenerator-runtime/runtime.js:63:40)
    at Generator.invoke [as _invoke] (/home/amir/Documents/code/reflog/node_modules/regenerator-runtime/runtime.js:294:22)
    at Generator.next (/home/amir/Documents/code/reflog/node_modules/regenerator-runtime/runtime.js:119:21)
    at asyncGeneratorStep (/home/amir/Documents/code/reflog/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
    at _next (/home/amir/Documents/code/reflog/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
    at /home/amir/Documents/code/reflog/node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
    at new Promise (<anonymous>)
    at /home/amir/Documents/code/reflog/node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12 {
  name: 'GetUserByAccountError',
  code: undefined
}
```
Now `GitHubProvider` wants to use the adapter which I don't want to happen. 

Also writing a custom adapter doesn't seem to solve my problem. I use typescript and the [`Adapter`](https://github.com/nextauthjs/next-auth/blob/83a9867cbcca95a50e4df3a6b8666ae8391db078/packages/next-auth/src/adapters.ts#L58) type still enforces the database structure that adapters expect.

Is it possible to do passwordless login without an adapter?

**Comment by alannnc** [comment]
Hey, we just run in to the same issue here? Do you happen to find a solution for this use case?

**Reply by amirhhashemi** [reply]
Hey
Sorry for the delayed response

> Do you happen to find a solution for this use case?

Not really. [This thread](https://github.com/nextauthjs/next-auth/discussions/4059#discussioncomment-2292120) might help you but I think it's too hacky for production. I managed to change my db schema and I suggest you do that too (if you really want this feature), otherwise wait for this feature to be officially supported (in an unknown time!!).

Let me know if you found a better solution.

**Reply by alannnc** [reply]
What I did ended up with, I made a replica adapter from prisma adapter from next-auth but I changed so it can use my custom fields for identifying user on my DB. Will work for a while as I setup a middle point for later update to fully use prisma adapter. I think this should work for you too, just change default methods behaviour. 

Do let me know too if there is something I can help you with.

https://github.com/calcom/cal.com/pull/2594/files#diff-09111951461813d687bb6933ed049950d934512d3c2758c0e98120f028d37875
